tower_cli/resources/setting.py

```diff
@@ -0,0 +1,108 @@
+# Copyright 2017, Ansible by Red Hat
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import click
+import six
+
+from tower_cli import models, resources
+from tower_cli.api import client
+from tower_cli.utils import exceptions as exc
+from tower_cli.utils.data_structures import OrderedDict
+from tower_cli.utils.decorators import pop_option
+
+
+class Resource(models.Resource):
+    cli_help = 'Manage settings within Ansible Tower.'
+    custom_category = None
+
+    value = models.Field(required=True)
+
+    @resources.command(ignore_defaults=True, no_args_is_help=False)
+    @click.option('category', '-c', '--category',
+                  help='If set, filter settings by a specific category')
+    def list(self, **kwargs):
+        """Return a list of objects."""
+        self.custom_category = kwargs.get('category', 'all')
+        try:
+            result = super(Resource, self).list(**kwargs)
+        except exc.NotFound as e:
+            categories = map(
+                lambda category: category['slug'],
+                client.get('/settings/').json()['results']
+            )
+            e.message = '%s is not a valid category.  Choose from [%s]' % (
+                kwargs['category'],
+                ', '.join(categories)
+            )
+            raise e
+        finally:
+            self.custom_category = None
+        return {
+            'results': [{'id': k, 'value': v} for k, v in result.items()]
+        }
+
+    @resources.command(ignore_defaults=True)
+    def get(self, pk, **kwargs):
+        """Return one and exactly one object"""
+        # The Tower API doesn't provide a mechanism for retrieving a single
+        # setting value at a time, so fetch them all and filter
+        try:
+            return next(s for s in self.list()['results'] if s['id'] == pk)
+        except StopIteration:
+            raise exc.NotFound('The requested object could not be found.')
```

Comment:
I checked out your branch and this works:

`tower-cli setting get SOCIAL_AUTH_SAML_ORGANIZATION_MAP`

This does not work:

`tower-cli setting get --value=SOCIAL_AUTH_SAML_ORGANIZATION_MAP`

That's fine, I somewhat prefer the one that works, but the help text:

```
$ tower-cli setting get
Usage: tower-cli setting get [OPTIONS] [ID]

  Return one and exactly one setting

Options:
  --value TEXT                    [REQUIRED] The value field.
  -h, --tower-host TEXT           The location of the Ansible Tower host.
                                  HTTPS is assumed as the protocol unless
                                  "http://" is explicitly provided. This will
                                  take precedence over a host provided to
                                  `tower config`, if any.
  -u, --tower-username TEXT       Username to use to authenticate to Ansible
                                  Tower. This will take precedence over a
                                  username provided to `tower config`, if any.
  -p, --tower-password TEXT       Password to use to authenticate to Ansible
                                  Tower. This will take precedence over a
                                  password provided to `tower config`, if any.
  -f, --format [human|json|yaml]  Output format. The "human" format is
                                  intended for humans reading output on the
                                  CLI; the "json" and "yaml" formats provide
                                  more data.
  -v, --verbose                   Show information about requests being made.
  --description-on                Show description in human-formatted output.
  --insecure                      Turn off insecure connection warnings. Set
                                  config verify_ssl to make this permanent.
  --certificate TEXT              Path to a custom certificate file that will
                                  be used throughout the command. Overwritten
                                  by --insecure flag if set.
  --help                          Show this message and exit.
```

I have a sense of what your edits to base.py did, and since the other standard use still works I'm fine with that, but you'll see that the help text points us to the `--value` usage. Because of that, use_fields_as_options may need to be revisited.